apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+                        var streamingPullCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
+                        pull = _client.StreamingPull(CallSettings.FromCancellationToken(streamingPullCts.Token),
+                            new BidirectionalStreamingSettings(10));// TODO: Put back to 1 once bug #188 released in v2.1.0
+                        // Initial call to start subscribe messages arriving.
+                        await _taskHelper.ConfigureAwait(pull.WriteAsync(new StreamingPullRequest
+                        {
+                            SubscriptionAsSubscriptionName = _subscriptionName,
+                            StreamAckDeadlineSeconds = _modifyDeadlineSeconds
+                        }));
+                        // Start Push task which pushes acks and extends.
+                        var pushTask = _taskHelper.Run(() => Push(pull, streamingPullCts.Token));
+                        // Read incoming messages whilst not stopped, and not in a fault condition.
+                        var pullTask = _taskHelper.Run(() => Pull(pull, streamingPullCts.Token));
+                        // Wait for either pull or push to complete.
+                        var completedTask = await _taskHelper.ConfigureAwait(_taskHelper.WhenAny(pushTask, pullTask));
+                        // Only cancel streaming if one of the tasks faulted. Otherwise use the non-abortive shutdown.
+                        if (completedTask.IsFaulted)
+                        {
+                            streamingPullCts.Cancel();
+                        }
+                        _qLock.Locked(() => _qEvent).TrySetResult(0);
+                        await _taskHelper.ConfigureAwait(_taskHelper.WhenAll(pushTask, pullTask));
+                    }
+                    catch (Exception e) when (e.IsCancellation() || e.IsRpcCancellation() || (e.As<RpcException>()?.IsRecoverable() ?? false))
+                    {
+                        // If it's not an unrecoverable error, then allow non-exceptional code-paths to handle. Do nothing here.
+                    }
+                    finally
+                    {
+                        // Try to cleanly shutdown the subscriber. This may or may not be possible
+                        // depending on the state of the subscriber; ignore all errors.
+                        if (pull != null)
+                        {
+                            _registerTaskFn(_taskHelper.Run(() => pull.WriteCompleteAsync()));
+                        }
+                    }
+                }
+            }
+
+            private async Task Pull(SubscriberClient.StreamingPullStream pull, CancellationToken streamingPullToken)
+            {
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(streamingPullToken, _softStopCts.Token);
+                while (!cts.IsCancellationRequested && await _taskHelper.ConfigureAwait(pull.ResponseStream.MoveNext(CancellationToken.None)))
```

Comment:
Why are we using CancellationToken.None here? Doesn't that mean we could be blocked waiting for the next item? Or is the point that if something gets cancelled, the response stream will finish?